### ISSUE TITLE
Override a css variable with the css-vars-ponyfill

### ISSUE THREAD

**Issue by nkatarow** [jhildenbiddle/css-vars-ponyfill#28]
Hi there, we're loving the ponyfill so far and are finding it super useful thus far. We're just running into one scenario, which is most likely an edge case, but one we'd like to account for and are hoping you might be able to advise on what might be possible.

Right now we're using the ponyfill as directed, defining our css variables in the :root selector, i.e.
```
:root {
  --button-bg-color: red;
}
```

We know trying to set the value of a css var anywhere else will be ignored by the polyfill ( but correctly processed by all browsers supporting css vars: anything but IE), i.e:
```
.my-special-button {
   --button-bg-color: blue
}
```

Given the two definitions above the goal would be to be able to do something along those lines:
```
<button></button> // this button is red on any browser
<button class="my-special-button"></button> // this button should be blue on any browser but right now will be red on IE because the value definition will be ignored

```
Would you have any recommendations on how we could manually set a var value on a specific HTML element through the ponyfill API?

**Comment by jhildenbiddle** [comment]
Hi @nkatarow.

Good to hear the ponyfill is working well for you!

The easiest option is to create multiple CSS custom properties and assign them to their own class. I prefer [BEM](http://getbem.com/introduction/)-style CSS selectors, but you can use whichever selector or class naming convention you prefer.

```css
:root {
  --button-bg-color: red;
  --button-bg-color--special: blue;
}

.button {
  /* Base styles here */
  border: 1px solid black;
  border-radius: 3px;
  background-color: var(--button-bg-color);
  color: white;
}

.button--special {
  /* Modifier styles here */
  background-color: var(--button-bg-color--special);
}
```

```html
<button class="button">Red Button</button>
<button class="button--special">Blue Button</button>
```

It's not as slick as using scoped custom properties, but it's arguably easier to manage because custom properties values remain grouped within `:root` selectors.

Hope this helps.

**Comment by nkatarow** [comment]
Unfortunately, this is not an option for us. The reason being is that our library will contain the :root and .button selectors and will be not be directly editable by the external user, so the idea would be that the external user needs to override those selectors from the shadow dom using something like:

`
.button--special {
  --button-bg-color: blue;
} 
`

With that said, would it be possible to declare css var not anywhere, but specifically on a shadow root? Maybe calling the ponyfill only on a subpart of the dom instead of reprocessing all? Or something along those lines?

**Comment by jhildenbiddle** [comment]
From what I can gather, it sounds like you are building a library the contains styles which you want consumers of the library to be able to override using CSS custom properties, yes?

If this is the case, then know that there are only two ways to override existing styles using the ponyfill:

1. **Declare override values in CSS by adding another `:root` ruleset**

   Your document can contain as many `:root` rulesets as you like, and the last value declared for a custom property wins. You can use this behavior to set your default values in one stylesheet, then instruct consumers of your library to specify custom property values in a separate stylesheet that appears *after* you library's stylesheet.

   **Library**

   ```css
   :root {
     --my-color: red;
   }

   button {
     color: var(--my-color);
   }
   ```

   **User** (added after **Library** via `<style>` or `<link>`)

   ```css
   :root {
     --my-color: blue;
   }
   ```

   **Result**
   ```css
   button {
     color: blue;
   }
   ```

   This is the method I have used in projects like [docsify-themeable](https://jhildenbiddle.github.io/docsify-themeable) and [docsify-tabs](https://jhildenbiddle.github.io/docsify-tabs) to allow for customization using CSS custom properties. See the "Customization" section of each project for details.

1. **Declare override values in JS using [`options.variables`](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsvariables)**

   The values provided will override any values declared within the document (see the [docs](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsvariables) for details).

   ```js
   cssVars({
     variables: {
       '--my-color': 'blue'
     }
   });
   ```

   Since you obviously don't want users calling `cssVars()` directly, you'll have to provide another means for them to pass an object containing name:value pairs to your lib. Once you have that object, just pass it directly to the ponyfill.

   ```js
   cssVars({
     variables: userLibSettings.variables
   });
   ```

I'm not sure if this solves your problem though because I'm not certain I understand the use case. If the above suggestions aren't what you're looking for, I can (hopefully) provide better assistance if I have better understanding of exactly what it is you're trying to do.

As far as declaring custom property values with the shadowDOM goes, the short answer is that this won't work because `:root` selectors are ignored in shadowDOM stylesheets the same way other selectors (like `body`) are ignored. This is by design and according to spec. The ponyfill *could* be modified to support this, but doing so would mean the results would be different in modern browsers (which ignore shadowDOM `:root` declarations) and legacy browsers (which could process shadowDOM `:root` declarations via the ponyfill) so this isn't really an option.

**Comment by nkatarow** [comment]
So using your example above, we'd still like the result to output the library specified color. However, in addition, we'd like the external user to be able to override that variable when necessary with a css class, only changing the instances containing that class. 

It's sounding like we'd just need to have them create a new variable or manually assign a value to the special/override class instead of being able to simply update the variable's value. Is that correct?

**Result**
`
.button { color: red; }
.button--special { color: blue; }
`

**Comment by jhildenbiddle** [comment]
Can you share more details about your library? 

I know your library provides default styles, works with shadow DOM in some fashion, and that you want to override default styles for specific elements. What isn't clear given the examples provided is the purpose of your library, how external users will use it, what limitations must be considered, and how the use of CSS custom properties provides value that standard CSS doesn't already offer.

For example, if the expectation is that external users will write their own CSS to override styles, how is this CSS...

```css
.my-special-button {
   --button-bg-color: blue;
}
```

... better than this css:

```css
.my-special-button {
  color: blue;
}
```

Happy to help and work towards a solution, but I'll need more details to do it.

**Comment by benoitjchevalier** [comment]
Hey @jhildenbiddle , I work with @nkatarow and hope I can help clarify what we're trying to achieve and potential solutions we're exploring.

We are building a library of webcomponents (a Design System) with Shadow Dom (using angular elements). Our library of components includes theming, which we achieve through the use of css variables to effectively "pierce" the shadow dom. Using your ponyfill we use the javascript entrypoint to set all our theming variables and switch between our themes (effectively setting all those variables on the <html> element for browsers with native support of css vars, and having the ponyfill doing the heavy work for the other).

This works fine and we're happy to have theming working on all browsers now.

The other feature that we need, both for our consumers and ourselves, is the ability to override one of the css variables on a specific component. The idea behind it is:
- the consumer needs to override the style of just one component for design reasons (e.g in a very specific case in their application the themed version doesn't work well and they need to change the border color. However other instances of this component in other places still need to be the vanilla version)
- the consumer or ourselves are building more complex components using simpler components. This more complex component need to override the styles of the simpler one to create a new visually compelling component. e.g: we create a datepicker which uses our dropdown component. The datepicker has 3 dropdowns next to each other and we want to remove the borders to make it look more like one unit.

The actual example I've given are just random thoughts so don't think too long on the design implication of what I'm saying, but the scenarios themselves are genuine problem we have to solve. The flexibility to override some styles of a specific component is a big helper in having a Design System adopted since people still have some control over those components and don't feel the need to fork them for specific scenarios.


Now one solution we thought of would be to provide a function on our components that would
- take an object with key-value pairs of css variables and their values
- take the `<style>` node of the component
- have the ponyfill process the whole string of the <style> node using those key-value with the `updateDom` option to false and accessing the resulting string
- append resulting string in a new `<style>` node in the Shadow Dom of the element

It's not ideal but might give us that override we're looking after. We're also open to any suggestion you might have on solving this, if you think there is a thing that would make sense to be baked in the ponyfill we can help as well :)

**Comment by jhildenbiddle** [comment]
Got it. Thanks for the detail, @benoitjchevalier!

To recap, your library will provide a collection of custom elements along with default styling for each.

```html
<custom-element>Hello!</custom-element> /* color: red (default) */
```

Each custom element contains a `<style>` element within its shadow Root that provides default styles and the ability to override using CSS custom properties.

```html
/* Viewed with Inspector */
<custom-element>
  #shadow-root
    <style>
      div {
        color: var(--my-color, red);
      }
    </style>
    <div>Hello!</div>
</custom-element>
```

For modern browsers with native CSS custom property support, customizing styles can by done in one of the following ways:

1. On an individual element using the `style` attribute:

   ```html
   <custom-element style="--my-color: blue;">Hello!</custom-element> /* color: blue */
   ```

1. On one-or-more elements using a CSS ruleset:

   ```css
   /* All */
   custom-element {
     --my-color: blue;
   }

   /* Some */
   .my-class {
     --my-color: green;
   }
   ```
   ```html
   <custom-element>Hello!</custom-element> /* color: blue */
   <custom-element class="my-class">Hello!</custom-element> /* color: green */
   ```

This all works dandy with the exception of legacy browsers. Sound about right?

---

#### TL;DR:

1. Handling individual elements seems doable.
1. Handling one-or-more elements using a CSS ruleset will be a nightmare unless you put strict limits on the type of rulesets you support and how they can be applied.

#### 1. Customizing individual element using the `style` attribute or JavaScript

This one is pretty straightforward: 

* Find all custom elements
* Parse each element's style attribute value for custom property declarations
* If a custom property exists, call the ponyfill with `options.variables` set to an object containing parsed name:value pairs and `rootElement` set to the custom element.

Here's the basic idea:

```html
<foo-element style="--my-color: blue;">Foo</foo-element>
<bar-element style="--my-color: green;">Bar</bar-element>
```
```javascript
const customElements = [...document.querySelectorAll('foo-element, bar-element')];

customElements.forEach(elm => {
  const styleVars = elm.getAttribute('style')
    .split(';')
    .map(decl => decl.trim())
    .filter(decl => /^--/.test(decl))
    .reduce((prev, current) => {
      const decl = current.split(':').map(val => val.trim());
      const prop = decl[0];
      const val = decl[1];
      
      prev[prop] = val;
      
      return prev;
  }, {});

   // 1. styleVars = { '--my-color': 'blue' } for <foo-element>
   // 2. styleVars = { '--my-color': 'green' } for <bar-element>

  if (Object.keys(styleVars).length) {
    cssVars({
      rootElement: elm,
      variables: styleVars
    });
  }
});
```

I like your idea of adding a method to all custom elements that allows passing CSS custom property name:value pairs as well. This allows custom property changes to be handled by the ponyfill, which ensures both modern and legacy browsers are updated. It also provides a convenient method of modifying multiple custom property values at once (something the native DOM methods do not provide).

#### 2. Customizing one-or-more elements using a CSS ruleset

This is a much bigger challenge due to the complexities introduced by the the cascade (the "C" in "CSS").

Simple examples make things seem... well, simple. Take the following for example:

```css
.my-class {
   --my-color: blue;
}
```
```html
<custom-element>Red</custom-element> /* color: red */
<custom-element class="my-class">Blue</custom-element> /* color: blue */
```

It seems relatively simple to figure out that the ".my-class" element should have "--my-color: blue" applied to it. This is more-or-less the example everyone uses when talking about scoped custom property support. The problem is that in addition to the simple example above, the following also should be supported:

```css
.my-class {
   --my-color: blue;
}
```
```html
<div class="my-class">
  <custom-element>Blue</custom-element> /* color: blue */
</div>
```

Now parent/child relationships have to be taken into consideration. 

This should also be supported:

```css
.foo:not(:first-child) {
   --my-color: blue;
}
.bar {
  --my-color: green;
}
.foo + .bar {
  --my-color: purple;
}
.baz {
  --my-color: pink;
}
.my-class custom-element:last-child {
  --my-color: cyan !important;
}
#buzz {
  --my-color: black;
}
```
```html
<div class="my-class">
  <custom-element class="foo">Red</custom-element> /* color: red (default) */
  <custom-element class="bar">Purple</custom-element> /* color: purple */
  <custom-element id="buzz" class="baz">Cyan</custom-element> /* color: cyan */
</div>
```

Yeesh. You get the idea.

Parent/child/sibling relationships, attributes, CSS specificity (when more than ruleset applies), media queries, etc. all have to be taken into account to determine the value(s) that should be applied to an element. Even if this was possible, it's a safe bet that the performance impact of recalculating the these styles each time a change to the cascade is made (e.g. a class name is added/removed, an element is inserted before/after, an element is enabled/disabled, etc.) would make the whole thing unsuitable for production.

This, in a nutshell, is why a CSS custom property poly/ponyfill with scoped property support doesn't exist.

The only option I can think of is to limit the type of rulesets you support, effectively preventing you from having to deal with the cascade. For example, you could specify that rulesets created for customizing your library's components *must* use one of the following formats:

```css
/* Tag (all instances) */
custom-element {
  --my-color: blue;
}

/* Single Class */
custom-element.my-class {
  --my-color: green;
}
```

Nothing else is supported. No IDs, nested classes, multi-class selectors, sibling selectors, attribute selectors, etc.

This works as-is in modern browsers, and makes parsing the CSS for tag + class name combinations much easier (you could even use the [parser](https://github.com/jhildenbiddle/css-vars-ponyfill/blob/master/src/parse-css.js) included in this ponyfill). When processing a `<custom-element>` object, you only need to look for rulesets with a `custom-element` or `custom-element.class-name` selector.

Keep in mind that you will still have to deal with multiple matching rulesets:

```css
custom-element{
  --my-color: blue;
}
custom-element.foo {
  --my-color: green;
}
custom-element.bar {
  --my-background: orange;
}
```
```html
<custom-element>Blue</custom-element> /* color: blue; */
<custom-element class="foo bar">Green on Orange</custom-element> /* color: green; background: orange; */
```

The other option is to just inform people that customization using class names and scoped custom properties simply isn't available for legacy browsers. Offering code samples demonstrating how to inject CSS into the shadow root of elements that need customization may be enough until legacy support is no longer required.

Whew. 😄 

Not sure if any of this solves your problem, but hopefully it moves the conversation forward a bit.

**Comment by benoitjchevalier** [comment]
Very detailed answer once again, much appreciated :)

I managed to spend some time on this this afternoon and think we have a way forward by using 2). We're going to focus on one solution that works, can be the sole mechanism used (both for modern and legacy browser) as well as letting people use the css variables in their css if they are not targeting IE.

My assumption was that I was going to be able to essentially write a small wrapper around the ponyfill, mostly leveraging the `rootElement` and `shadowDOM` options, but I'm getting some unexpected results so wanted to clarify my assumptions with you.

I'm assuming that `rootElement` would be both the element from where the styles are being read as well as the "starting point" from where the polyfill would do the work. The exception being in `shadowDOM` mode with the ponyfill reading the css vars values at the root (which is helpful).

However when trying to apply a single value on a custom element I can see it being applied:
- to all elements in the page on Chrome. Stop me if I'm wrong but shouldn't [this line](https://github.com/jhildenbiddle/css-vars-ponyfill/blob/master/src/index.js#L208) apply the prop to the root element instead of the document? i.e

from
```
```

**Comment by benoitjchevalier** [comment]
Very detailed answer once again, much appreciated :)

I managed to spend some time on this this afternoon and think we have a way forward by using 2). We're going to focus on one solution that works, can be the sole mechanism used (both for modern and legacy browser) as well as letting people use the css variables in their css if they are not targeting IE.

My assumption was that I was going to be able to essentially write a small wrapper around the ponyfill, mostly leveraging the `rootElement` and `shadowDOM` options, but I'm getting some unexpected results so wanted to clarify my assumptions with you.

I'm assuming that `rootElement` would be both the element from where the styles are being read as well as the "starting point" from where the polyfill would do the work. The exception being in `shadowDOM` mode with the ponyfill reading the css vars values at the root (which is helpful).

However when trying to apply a single value on a custom element I can see it being applied everywhere that value is used in the page on Chrome. Stop me if I'm wrong but shouldn't [this line](https://github.com/jhildenbiddle/css-vars-ponyfill/blob/master/src/index.js#L208) apply the prop to the rootElement instead of the document? i.e

from
```
document.documentElement.style.setProperty(prop, value);
```

to something like that
```
var target = settings.rootElement === document ? document.documentElement : settings.rootElement.host ? settings.rootElement.host : settings.rootElement;
target.style.setProperty(prop, value);
```

I'm also seeing issue applying a single value on a custom element in IE (it's applied everywhere) but after some debugging it looks like the ponyfill is doing the right thing and my problem lies with shadyCSS in angular elements

**Comment by jhildenbiddle** [comment]
Good catch, @benoitjchevalier.

> However when trying to apply a single value on a custom element I can see it being applied everywhere that value is used in the page on Chrome.

This bug has been fixed in 1.11.3.

As for the IE/shadyCSS issue, I'm not sure how much help I can provide other than to suggest a path for narrowing down the issue:

1. Creating a basic custom element + ponyfill setup to confirm that the ponyfill is behaving as expected. I've created a [CodePen](https://codepen.io/jhildenbiddle/pen/WaaLXL) to help get you started (modern browsers only).
1.  Introduce web component polyfills to the CodePen demo to get it working in legacy browsers and verify if shadowCSS is the issue. Unfortunately CodePen doesn't support IE11, but you can view a read-only version of the Pen in IE11 by manually changing to "Full Page" view. 

Give me a shout if there's anything else I can do to help! 😃

**Comment by benoitjchevalier** [comment]
Ace 👍 . I've taken my second issue to the Angular guys, I verified that the ponyfill is working as expected by debugging on IE, the problem is around Angular Element not providing style encapsulation for browsers not supporting Shadow Dom.

Feel free to close this issue and thanks for your time!

**Comment by jhildenbiddle** [comment]
Happy to help, and thank you for bringing the bug fixed in 1.11.3 to my attention.

Best of luck with your project!
